### PR TITLE
Batch trie keys

### PR DESCRIPTION
**Pull request by matkt**
## PR description

Introduce a TrieKeyBatchAdapter to simultaneously generate multiple trie keys, thereby optimizing performance.

Fixing #45

### CODE REVIEW

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java [hunk 1]
```diff
@@ -53,16 +53,8 @@ public TrieKeyAdapter(Hasher hasher) {
...
-   * Swaps the last byte of the base key with a given subIndex.
-   *
-   * @param base The base key.
-   * @param subIndex The subIndex.
-   * @return The modified key.
-   */
-  Bytes32 swapLastByte(Bytes32 base, UInt256 subIndex) {
-    Bytes lastByte = Bytes.of(subIndex.toBytes().reverse().get(0));
-    return (Bytes32) Bytes.concatenate(base.slice(0, 31), lastByte);
+  public Hasher getHasher() {
```
Comment (garyschulte) [inline src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java:56]: nit: javadoc

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/hasher/PedersenHasher.java [hunk 2]
```diff
@@ -82,6 +88,56 @@ public Bytes32 groupToField(Bytes input) {
...
+
+    final Bytes hash =
+        Bytes.wrap(
+            LibIpaMultipoint.hash(LibIpaMultipoint.commit(Bytes.concatenate(chunks).toArray())));
+
+    // commitRoot returns the hash in big endian format, so we reverse it to get it
+    // in little endian
+    // format. When we migrate to using `groupToField`, this reverse will not be
+    // needed.
+    return Bytes32.wrap(hash);
```
Comment (garyschulte) [inline src/main/java/org/hyperledger/besu/ethereum/trie/verkle/hasher/PedersenHasher.java:104]: given the comment, should this be `hash.reverse()` ?

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java [hunk 3]
```diff
@@ -73,25 +65,18 @@ Bytes32 swapLastByte(Bytes32 base, UInt256 subIndex) {
...
-   *
-   * @param address The address.
-   * @param chunkId The chunk ID.
-   * @return The generated code chunk key.
-   */
-  public Bytes32 codeChunkKey(Bytes address, int chunkId) {
-    return codeChunkKey(address, UInt256.valueOf(chunkId));
+  protected UInt256 locateStorageKeyOffset(Bytes32 storageKey) {
+    UInt256 index = UInt256.fromBytes(storageKey);
+    UInt256 headerOffset = CODE_OFFSET.subtract(HEADER_STORAGE_OFFSET);
```
[the remark this answers is not in the document]
Comment (gfukushima) [reply src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java:76]: nice spot, easy optimization

File: src/test/java/org/hyperledger/besu/ethereum/trie/verkle/TrieKeyBatchAdapterTest.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            .boxed()
+            .collect(Collectors.toMap(expectedKeys::get, expectedValues::get));
+    final Map<Bytes32, Bytes32> generatedHashes =
+        adapter.manyAccountTrieKeys(address, expectedKeys, new ArrayList<>(), new ArrayList<>());
+    assertThat(generatedHashes).containsExactlyEntriesOf(map);
+  }
+
+  @Test
+  public void testAccountKeysWithStorage() {
+    final List<Bytes32> expectedKeys = new ArrayList<>();
```
Comment (gfukushima) [inline src/test/java/org/hyperledger/besu/ethereum/trie/verkle/TrieKeyBatchAdapterTest.java:77]: I think these are the expected indexes we want to generate keys for?
```suggestion
    final List<Bytes32> expectedIndexes = new ArrayList<>();
```

File: src/test/java/org/hyperledger/besu/ethereum/trie/verkle/TrieKeyBatchAdapterTest.java [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    expectedKeys.add(Parameters.VERSION_LEAF_KEY);
+    expectedKeys.add(Parameters.BALANCE_LEAF_KEY);
+    expectedKeys.add(Parameters.NONCE_LEAF_KEY);
+    expectedKeys.add(Parameters.CODE_KECCAK_LEAF_KEY);
+    expectedKeys.add(Parameters.CODE_SIZE_LEAF_KEY);
+
+    final UInt256 storage = UInt256.valueOf(64);
+    expectedKeys.add(storage);
+
+    final List<Bytes32> expectedValues =
```
Comment (gfukushima) [inline src/test/java/org/hyperledger/besu/ethereum/trie/verkle/TrieKeyBatchAdapterTest.java:87]: think these are the expected keys, just a rename for a better comprehension
```suggestion
    final List<Bytes32> expectedKeys =
```

**Review by gfukushima — APPROVED** [review]
lgtm, some nice to have spotted by Gary if you could that add to this PR already that would be awesome.

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java (hunk 1 above)
Comment (matkt) [reply src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java:56]: done

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (matkt) [reply src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java:76]: good catch. updated the code

File: src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java [hunk 6]
```diff
@@ -117,8 +105,19 @@ public Bytes32 codeChunkKey(Bytes address, UInt256 chunkId) {
...
+
+  /**
+   * Swaps the last byte of the base key with a given subIndex.
+   *
+   * @param base The base key.
+   * @param subIndex The subIndex.
+   * @return The modified key.
+   */
+  Bytes32 swapLastByte(Bytes32 base, Bytes32 subIndex) {
+    Bytes lastByte = Bytes.of(subIndex.reverse().get(0));
```
[the remark this answers is not in the document]
Comment (matkt) [reply src/main/java/org/hyperledger/besu/ethereum/trie/verkle/adapter/TrieKeyAdapter.java:119]: aaah yes clear better thanks

File: src/test/java/org/hyperledger/besu/ethereum/trie/verkle/TrieKeyBatchAdapterTest.java (hunk 5 above)
Comment (matkt) [reply src/test/java/org/hyperledger/besu/ethereum/trie/verkle/TrieKeyBatchAdapterTest.java:87]: done

**Review by garyschulte — APPROVED** [review]
LGTM 👍

**Pull request merged**